### PR TITLE
Projects controller #40

### PR DESCRIPTION
**Pull request by pospieszynski**
projects controller and corresponding specs initial version
feedback will be appreciated;]

### CODE REVIEW

File: spec/controllers/projects_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+      before { get :index, params: { id: project.id } }
+
+      it "returns http success" do
+        expect(response).to have_http_status(:success)
+      end
+
+      it "assigns all tasks to the instance variable" do
+        expect(assigns(:projects)).to eq(Project.all)
+      end
+      it "renders template index" do
```
Comment (tprochniewicz) [inline spec/controllers/projects_controller_spec.rb]: empty line above

File: spec/controllers/projects_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+    describe "GET #index" do
+      before { get :index, params: { id: project.id } }
+
+      it "returns http success" do
+        expect(response).to have_http_status(:success)
+      end
+
+      it "assigns all tasks to the instance variable" do
+        expect(assigns(:projects)).to eq(Project.all)
```
Comment (tprochniewicz) [inline spec/controllers/projects_controller_spec.rb]: in description you have 'all tasks'

File: spec/controllers/projects_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    end
+
+    describe "GET #managed_projects" do
+      before { get :managed_projects }
+
+      it "returns http success" do
+        expect(response).to have_http_status(:success)
+      end
+
+      it "assigns all tasks to the instance variable" do
```
Comment (tprochniewicz) [inline spec/controllers/projects_controller_spec.rb]: tasks / projects?

File: spec/controllers/projects_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+
+    describe "GET #contributed_projects" do
+      before { get :contributed_projects }
+
+      it "returns http success" do
+        expect(response).to have_http_status(:success)
+      end
+
+      it "assigns all tasks to the instance variable" do
```
Comment (tprochniewicz) [inline spec/controllers/projects_controller_spec.rb]: tasks / projects?

File: spec/controllers/projects_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+      it "redirects to projects url" do
+        expect(response).to redirect_to projects_path
+      end
+
+      it "deletes the requested record " do
+        expect(Project.all).to be_empty
+      end
+
+      it "displays roper flush message" do
```
Comment (tprochniewicz) [inline spec/controllers/projects_controller_spec.rb]: typo

File: spec/controllers/tasks_controller_spec.rb [hunk 6]
```diff
@@ -371,7 +371,6 @@
           end
 
           it "returns proper flush message" do
-            expect(flash[:notice]).to include("Error, requested task is not associated with this project")
```
Comment (tprochniewicz) [inline spec/controllers/tasks_controller_spec.rb]: ?

File: app/controllers/projects_controller.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  def show
+  end
+
+  def new
+    @project = current_user.managed_projects.new
+  end
+
+  def create
+    @project = current_user.managed_projects.new(project_params)
+    if @project.save && UserProject.create(user: current_user, project: @project)
```
Comment (smt116) [inline app/controllers/projects_controller.rb]: What if `@project.save` will succeed and `UserProject::create` will fail?

File: app/controllers/projects_controller.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  def set_project
+    @project = Project.find(params[:id])
+  end
+
+  def project_params
+    params.require(:project).permit(:title, :description, :date)
+  end
+
+  def record_not_found
+    render file: 'public/404.html', status: :not_found
```
Comment (smt116) [inline app/controllers/projects_controller.rb]: It is okay for now, but generally it should be handled somewhere higher to handle all 404 errors.

Comment (pospieszynski) [inline app/controllers/projects_controller.rb]: like in application_controller?

File: app/controllers/projects_controller.rb (hunk 7 above)
Comment (pospieszynski) [inline app/controllers/projects_controller.rb]: i will wrap in in transaction then.

File: app/controllers/projects_controller.rb (hunk 8 above)
Comment (smt116) [inline app/controllers/projects_controller.rb]: Yes.

File: app/controllers/projects_controller.rb (hunk 7 above)
Comment (pospieszynski) [inline app/controllers/projects_controller.rb]: in custom create method defined on the model layer

**Comment by tprochniewicz** [comment]
:+1: 
@gotar

File: app/controllers/projects_controller.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  def show
+  end
+
+  def new
+    @project = current_user.managed_projects.new
+  end
+
+  def create
+    @project = current_user.managed_projects.new(project_params)
+    if Project.create_project(current_user, @project)
```
Comment (gotar) [inline app/controllers/projects_controller.rb]: it should be sth like current_user.project.create(project_params)

File: app/models/project.rb [hunk 10]
```diff
@@ -8,4 +8,15 @@ class Project < ApplicationRecord
   has_many :events, dependent: :destroy
 
   has_many :tasks, dependent: :destroy
+
+  def self.create_project(current_user, project)
+    transaction do
+      begin
+        project.save
+        UserProject.create(user: current_user, project: project)
```
Comment (gotar) [inline app/models/project.rb]: this will be done automatically if you have proper relation and call it in proper way

Comment (gotar) [inline app/models/project.rb]: but +1 for idea and transaction, good job

**Comment by gotar** [comment]
good enough I put some comments, depend of you - merge, or fix and then merge (remember to squash)

**Pull request merged**